File: dojo/mik.s/RomanCalculator/CSharp/2012-06-27 Yellow Belt/RomanCalculator/Calculator.cs
using System.Linq;
using System.Text;
using RomanCalculator.Interfaces;

namespace RomanCalculator
{
    public class Calculator : ICalculator
    {
        public string Add(string n1, string n2)
        {
            n1 = RemoveSubtractions(n1);
            n2 = RemoveSubtractions(n2);

            var sum = n1 + n2;
            sum = SortNumerals(sum);
            sum = CollapseRepeats(sum);
            sum = ApplySubtractions(sum);

            return sum;
        }

        #region Private helpers

        private readonly char[] _numerals = new [] { 'M', 'D', 'C', 'L', 'X', 'V', 'I'};

        private readonly string[][] _subtractions = 
            new [] {
                new [] {"IV", "IIII"},
                new [] {"IX", "VIIII"},
                new [] {"XL", "XXXX"},
                new [] {"XC", "LXXXX"}, 
                new [] {"CD", "CCCC"},
                new [] {"CM", "DCCCC"}
            };

        private readonly string[][] _repeats =
            new [] {
                new[] {"IIIII", "V"},
                new[] {"VV", "X"},
                new[] {"XXXXX", "L"},
                new[] {"LL", "C"},
                new[] {"CCCCC", "D"},
                new[] {"DD", "M"}
            };

        private string RemoveSubtractions(string n)
        {
            return _subtractions.Aggregate(n, (current, pair) => current.Replace(pair[0], pair[1]));
        }

        private string CollapseRepeats(string n)
        {
            return _repeats.Aggregate(n, (current, pair) => current.Replace(pair[0], pair[1]));
        }

        private string ApplySubtractions(string n)
        {
            return _subtractions.Aggregate(n, (current, pair) => current.Replace(pair[1], pair[0]));
        }

        private string SortNumerals(string n)
        {
            // I considered doing this via a custom IComparer, but that would likely have been more code and 
            // would likely not have performed any better on small strings like these.
            var sb = new StringBuilder();

            foreach (var numeral in _numerals)
            {
                foreach (var c in n)
                {
                    if (c == numeral)
                    {
                        sb.Append(c);
                    }
                }
            }

            return sb.ToString();
        }

        #endregion
    }
}

File: dojo/mik.s/RomanCalculator/Javascript/2012-06-27 Yellow Belt/src/Calculator.js
function Calculator() {
}

// Adding to the prototype instead of individual objects (in the constructor 
// above) conserves memory at the slight expense of lookup delay
Calculator.prototype.numerals = [
    "M", "D", "C", "L", "X", "V", "I"
];

Calculator.prototype.subtractions = [
    ["IV", "IIII"],
    ["IX", "VIIII"],
    ["XL", "XXXX"],
    ["XC", "LXXXX"],
    ["CD", "CCCC"],
    ["CM", "DCCCC"]
];

Calculator.prototype.repeats = [
    ["IIIII", "V"],
    ["VV", "X"],
    ["XXXXX", "L"],
    ["LL", "C"],
    ["CCCCC", "D"],
    ["DD", "M"]
];

Calculator.prototype.multiReplace = function(n, replacements, reverse) {
    var text = n;
    
    $(replacements).each(function() {
        text = text.replace(
            new RegExp(this[reverse ? 1 : 0], "g"), 
            this[reverse ? 0 : 1]
        );
    });
    
    return text;
};

Calculator.prototype.removeSubtractions = function(n) {
    return this.multiReplace(n, this.subtractions);
};

Calculator.prototype.collapseRepeats = function(n) {
    return this.multiReplace(n, this.repeats);
};

Calculator.prototype.applySubtractions = function(n) {
    return this.multiReplace(n, this.subtractions, true);
};

// There's probably a better way to do this in Javascript
Calculator.prototype.sortNumerals = function(n) {
    var result = "";
    var chars = n.split('');
    
    $(this.numerals).each(function() {
        var num = "" + this;        // Convert to actual string
        
        $(chars).each(function() {
            var ch = "" + this;     // Convert to actual string
            
            if (ch === num) {
                result = result + this;
            }
        });
    });
   
    return result;
};

Calculator.prototype.add = function(n1, n2) {	
	n1 = this.removeSubtractions(n1);
    n2 = this.removeSubtractions(n2);
    
    var sum = n1 + n2;
    sum = this.sortNumerals(sum);
    sum = this.collapseRepeats(sum);
    sum = this.applySubtractions(sum);

    return sum;
}